documentation/add/04_mpi_framework.md

```diff
@@ -1,43 +1,362 @@
+
+#### Common initialization.
+
+Different frameworks launch different types of parallel processes but each process should know what sub-task it needs to perform. The job description can be very different in size, so it is better to provide such description by writing appropriate description files. Command line arguments are used to provide each parallel worker with information about the location of a folder, where the job initialization information is placed on a shared file system plus some auxiliary information. Currently auxiliary information contains the name of the *message framework* used for communications between workers and worker number (*labNum*) for file-based messages framework. To avoid issues when different frameworks and different operating systems processing non-ASCII characters in a command line differently, the location information is encoded into single ASCII-128 string using standard Java base64 encoder. The *build_worker_init* method of **iMessagesFramework** performs the encoding of the input initialization information for further transfer of this information to parallel workers using command line arguments of correspondent job. 
+
+#### Interrupts or Persistent messages.
+Independent workers may need to report to the other workers some special conditions occurring during their execution. Such conditions currently is a failure (an interrupt is risen at code execution). We use Persistent messages aka Interrupt messages to distribute information about such states of the system as parallel interrupts, as MPI interrupts used for these purposes in standard MPI frameworks are not available in all parallel frameworks. Such message, received by a worker from other worker, block all future communications with this worker. All requests for information from this worker result in the persistent message (*failure*). This state holds until special command (*clear_all* above) is executed by the receiving worker.
+
+#### Messages types.
+
+Different purposes can be best served using different types of messages. There are 3 purposes, differentiated in Horace.
+
+    1) logging, diagnostics and informing user about progress of a job. 
+    2) Interprocess data exchange, necessary for performing a particular user job and 
+    3) The information about error, failure or job completion. 
+    
+Correspondent types of messages created to serve these purposes:
+
+    1) Non-blocking (transient)  messages. Informing user about a job progress is important task, but user is interested only in the final value. Different independent task should not be synchronous with regard to log messages as only some average progress is required. Non-blocking messages are best suited for such kind of task. If more then one non-blocking message is present in the MPI messages queue, the client receives only the last message.
```

Comment:
formatting -- the tab has turned these long lines into a horizontally scrolling code block that is hard to read

Reply:
fixed